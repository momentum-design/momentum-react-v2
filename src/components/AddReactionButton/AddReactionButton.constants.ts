const CLASS_PREFIX = 'md-add-reaction-button';

// setting to 24, cause momentum-design doesn't support size 20 for variant secondary
// overridding it in css to size 20
const DEFAULTS = { SIZE: 24 };

const STYLE = {
  wrapper: `${CLASS_PREFIX}-wrapper`,
};

export { CLASS_PREFIX, DEFAULTS, STYLE };
