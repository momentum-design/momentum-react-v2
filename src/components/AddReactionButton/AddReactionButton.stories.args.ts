import { commonMdcButton, commonStyles } from '../../storybook/helper.stories.argtypes';

export default {
  ...commonStyles,
  ...commonMdcButton,
};
