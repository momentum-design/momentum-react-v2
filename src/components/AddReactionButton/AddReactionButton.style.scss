// color overrides, cause Design specs deviate from Momentum Design specs
.md-add-reaction-button-wrapper {
  height: var(--mdc-button-height-size-20);
  margin: 0 0.125rem;
  background-color: var(--mds-color-theme-background-solid-primary-normal);

  .reaction-badge-count {
    margin-left: 0.25rem;
  }

  &:hover {
    background-color: var(--mds-color-theme-background-solid-secondary-normal);
  }

  &:active {
    background-color: var(--mds-color-theme-background-solid-tertiary-normal);
  }
}

.md-add-reaction-button-wrapper::part(prefix-icon) {
  --mdc-icon-size: 0.75rem;
}
