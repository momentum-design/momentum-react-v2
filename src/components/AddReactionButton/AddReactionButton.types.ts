import type { MdcButtonProps } from '../../types';

export type Props = MdcButtonProps;
