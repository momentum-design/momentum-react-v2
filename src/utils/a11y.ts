import { AriaAttributes, useEffect } from 'react';
import { RequireOneOf } from './types';

export type AriaLabelRequired = RequireOneOf<AriaAttributes, ['aria-label', 'aria-labelledby']>;

/**
 * Check if `aria-label` or `aria-labelledby` are defined and are truthy
 *
 * @param componentName The name of the component
 * @param props The object of props passed to the component
 */
export function useCheckAriaLabel(
  componentName: string,
  props: { 'aria-label'?: string; 'aria-labelledby'?: string }
): void {
  useEffect(() => {
    if (!props['aria-label'] && !props['aria-labelledby']) {
      console.warn(`MRV2: ${componentName} requires aria-label or aria-labelledby.`);
    }
  }, [componentName, props]);
}
