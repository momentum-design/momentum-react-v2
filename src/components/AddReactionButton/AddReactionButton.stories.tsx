import React from 'react';
import { MultiTemplate, Template } from '../../storybook/helper.stories.templates';
import { DocumentationPage } from '../../storybook/helper.stories.docs';
import StyleDocs from '../../storybook/docs.stories.style.mdx';

import AddReactionButton, { AddReactionButtonProps } from './';
import argTypes from './AddReactionButton.stories.args';
import Documentation from './AddReactionButton.stories.docs.mdx';
import TooltipPopoverCombo from '../TooltipPopoverCombo';

export default {
  title: 'Momentum UI/AddReactionButton',
  component: AddReactionButton,
  parameters: {
    expanded: true,
    docs: {
      page: DocumentationPage(Documentation, StyleDocs),
    },
  },
};

const Example = Template<AddReactionButtonProps>(AddReactionButton).bind({});

Example.argTypes = { ...argTypes };

Example.args = {};

const WithTooltipPopoverCombo = Template<AddReactionButtonProps>((args: any) => {
  return (
    <TooltipPopoverCombo
      otherTooltipProps={{
        placement: 'bottom',
        strategy: 'fixed',
      }}
      popoverContent={
        <div>
          <p>Popover content</p>
          <button>Button inside</button>
        </div>
      }
      tooltipContent={<p>Tooltip</p>}
      triggerComponent={<AddReactionButton {...args} />}
    />
  );
}).bind({});

Example.argTypes = { ...argTypes };

Example.args = {};

const Common = MultiTemplate<AddReactionButtonProps>(AddReactionButton).bind({});

Common.argTypes = { ...argTypes };
delete Common.argTypes.children;

Common.args = {};

Common.parameters = {
  variants: [{}, { disabled: true }, {}],
};

export { Example, WithTooltipPopoverCombo, Common };
