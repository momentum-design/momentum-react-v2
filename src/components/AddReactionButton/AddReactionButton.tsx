import React, { ForwardedRef, forwardRef } from 'react';
import classnames from 'classnames';
import ButtonCircle, { ButtonCircleSize } from '../ButtonCircle';

import { DEFAULTS, STYLE } from './AddReactionButton.constants';
import { Props } from './AddReactionButton.types';
import './AddReactionButton.style.scss';
import Icon, { IconScale } from '../Icon';

const AddReactionButton = forwardRef((props: Props, ref: ForwardedRef<HTMLButtonElement>) => {
  const { className, id, style, ...otherProps } = props;
  delete otherProps.size;
  return (
    <ButtonCircle
      ref={ref}
      className={classnames(className, STYLE.wrapper)}
      id={id}
      size={DEFAULTS.SIZE as ButtonCircleSize}
      style={style}
      {...otherProps}
    >
      <Icon name="reactions" scale={DEFAULTS.ICON_SIZE as IconScale} />
    </ButtonCircle>
  );
});

AddReactionButton.displayName = 'AddReactionButton';

export default AddReactionButton;
