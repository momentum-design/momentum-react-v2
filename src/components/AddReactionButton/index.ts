import { default as AddReactionButton } from './AddReactionButton';
import * as CONSTANTS from './AddReactionButton.constants';
import type { Props } from './AddReactionButton.types';

export { CONSTANTS as ADD_REACTION_BUTTON_CONSTANTS };

export type AddReactionButtonProps = Props;

export default AddReactionButton;
