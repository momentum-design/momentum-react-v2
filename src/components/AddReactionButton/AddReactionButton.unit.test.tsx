import React from 'react';
import { renderWithWebComponent } from '../../../test/utils';
import AddReactionButton, { ADD_REACTION_BUTTON_CONSTANTS as CONSTANTS } from './';
import userEvent from '@testing-library/user-event';
import { screen } from '@testing-library/react';

describe('<AddReactionButton />', () => {
  const setup = async (component: React.ReactElement) => {
    return renderWithWebComponent(component);
  };

  describe('snapshot', () => {
    it('should match snapshot', async () => {
      expect.assertions(1);

      const { container } = await setup(<AddReactionButton />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with className', async () => {
      expect.assertions(1);

      const className = 'example-class';

      const { container } = await setup(<AddReactionButton className={className} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with id', async () => {
      expect.assertions(1);

      const id = 'example-id';

      const { container } = await setup(<AddReactionButton id={id} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with style', async () => {
      expect.assertions(1);

      const style = { color: 'pink' };

      const { container } = await setup(<AddReactionButton style={style} />);

      expect(container).toMatchSnapshot();
    });
  });

  describe('attributes', () => {
    it('should have its wrapper class', async () => {
      expect.assertions(1);

      await setup(<AddReactionButton />);
      const element = screen.getByRole('button');

      expect(element.classList.contains(CONSTANTS.STYLE.wrapper)).toBe(true);
    });

    it('should have provided class when className is provided', async () => {
      expect.assertions(1);

      const className = 'example-class';

      await setup(<AddReactionButton className={className} />);
      const element = screen.getByRole('button');

      expect(element.classList.contains(className)).toBe(true);
    });

    it('should have provided id when id is provided', async () => {
      expect.assertions(1);

      const id = 'example-id';

      await setup(<AddReactionButton id={id} />);
      const element = screen.getByRole('button');

      expect(element.id).toBe(id);
    });

    it('should have provided style when style is provided', async () => {
      expect.assertions(1);

      const style = { color: 'pink' };
      const styleString = 'color: pink;';

      await setup(<AddReactionButton style={style} />);
      const element = screen.getByRole('button');

      expect(element.getAttribute('style')).toBe(styleString);
    });
  });

  describe('actions', () => {
    const user = userEvent.setup();

    it('should handle mouse click events', async () => {
      expect.assertions(1);

      const mockCallback = jest.fn();

      await setup(<AddReactionButton onClick={mockCallback} />);

      const component = screen.getByRole('button');

      await user.click(component);

      expect(mockCallback).toBeCalledTimes(1);
    });
  });
});
