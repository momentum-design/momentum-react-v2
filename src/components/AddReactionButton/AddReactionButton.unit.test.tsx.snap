// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddReactionButton /> snapshot should match snapshot 1`] = `
<AddReactionButton>
  <ButtonCircle
    className="md-add-reaction-button-wrapper"
    size={20}
  >
    <ButtonSimple
      aria-disabled={false}
      className="md-button-circle-wrapper md-add-reaction-button-wrapper"
      data-color="primary"
      data-disabled={false}
      data-ghost={false}
      data-inverted={false}
      data-multiple-children={false}
      data-outline={false}
      data-shallow-disabled={false}
      data-size={20}
    >
      <FocusRing>
        <FocusRing
          focusClass="md-focus-ring-wrapper children"
          focusRingClass="md-focus-ring-wrapper children"
        >
          <button
            className="md-button-circle-wrapper md-add-reaction-button-wrapper md-button-simple-wrapper"
            data-color="primary"
            data-disabled={false}
            data-ghost={false}
            data-inverted={false}
            data-multiple-children={false}
            data-outline={false}
            data-shallow-disabled={false}
            data-size={20}
            onBlur={[Function]}
            onClick={[Function]}
            onDragStart={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchCancel={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            type="button"
          >
            <Icon
              name="reactions"
              scale={12}
            >
              <div
                aria-hidden="true"
                className="md-icon-wrapper md-icon-auto-scales md-icon-scales md-icon-no-shrink"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  className=""
                  data-autoscale={false}
                  data-scale={12}
                  data-test="reactions"
                  fill="currentColor"
                  height="100%"
                  style={{}}
                  viewBox="0, 0, 32, 32"
                  width="100%"
                />
              </div>
            </Icon>
          </button>
        </FocusRing>
      </FocusRing>
    </ButtonSimple>
  </ButtonCircle>
</AddReactionButton>
`;

exports[`<AddReactionButton /> snapshot should match snapshot with className 1`] = `
<AddReactionButton
  className="example-class"
>
  <ButtonCircle
    className="example-class md-add-reaction-button-wrapper"
    size={20}
  >
    <ButtonSimple
      aria-disabled={false}
      className="md-button-circle-wrapper example-class md-add-reaction-button-wrapper"
      data-color="primary"
      data-disabled={false}
      data-ghost={false}
      data-inverted={false}
      data-multiple-children={false}
      data-outline={false}
      data-shallow-disabled={false}
      data-size={20}
    >
      <FocusRing>
        <FocusRing
          focusClass="md-focus-ring-wrapper children"
          focusRingClass="md-focus-ring-wrapper children"
        >
          <button
            className="md-button-circle-wrapper example-class md-add-reaction-button-wrapper md-button-simple-wrapper"
            data-color="primary"
            data-disabled={false}
            data-ghost={false}
            data-inverted={false}
            data-multiple-children={false}
            data-outline={false}
            data-shallow-disabled={false}
            data-size={20}
            onBlur={[Function]}
            onClick={[Function]}
            onDragStart={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchCancel={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            type="button"
          >
            <Icon
              name="reactions"
              scale={12}
            >
              <div
                aria-hidden="true"
                className="md-icon-wrapper md-icon-auto-scales md-icon-scales md-icon-no-shrink"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  className=""
                  data-autoscale={false}
                  data-scale={12}
                  data-test="reactions"
                  fill="currentColor"
                  height="100%"
                  style={{}}
                  viewBox="0, 0, 32, 32"
                  width="100%"
                />
              </div>
            </Icon>
          </button>
        </FocusRing>
      </FocusRing>
    </ButtonSimple>
  </ButtonCircle>
</AddReactionButton>
`;

exports[`<AddReactionButton /> snapshot should match snapshot with id 1`] = `
<AddReactionButton
  id="example-id"
>
  <ButtonCircle
    className="md-add-reaction-button-wrapper"
    id="example-id"
    size={20}
  >
    <ButtonSimple
      aria-disabled={false}
      className="md-button-circle-wrapper md-add-reaction-button-wrapper"
      data-color="primary"
      data-disabled={false}
      data-ghost={false}
      data-inverted={false}
      data-multiple-children={false}
      data-outline={false}
      data-shallow-disabled={false}
      data-size={20}
      id="example-id"
    >
      <FocusRing>
        <FocusRing
          focusClass="md-focus-ring-wrapper children"
          focusRingClass="md-focus-ring-wrapper children"
        >
          <button
            className="md-button-circle-wrapper md-add-reaction-button-wrapper md-button-simple-wrapper"
            data-color="primary"
            data-disabled={false}
            data-ghost={false}
            data-inverted={false}
            data-multiple-children={false}
            data-outline={false}
            data-shallow-disabled={false}
            data-size={20}
            id="example-id"
            onBlur={[Function]}
            onClick={[Function]}
            onDragStart={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchCancel={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            type="button"
          >
            <Icon
              name="reactions"
              scale={12}
            >
              <div
                aria-hidden="true"
                className="md-icon-wrapper md-icon-auto-scales md-icon-scales md-icon-no-shrink"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  className=""
                  data-autoscale={false}
                  data-scale={12}
                  data-test="reactions"
                  fill="currentColor"
                  height="100%"
                  style={{}}
                  viewBox="0, 0, 32, 32"
                  width="100%"
                />
              </div>
            </Icon>
          </button>
        </FocusRing>
      </FocusRing>
    </ButtonSimple>
  </ButtonCircle>
</AddReactionButton>
`;

exports[`<AddReactionButton /> snapshot should match snapshot with style 1`] = `
<AddReactionButton
  style={
    {
      "color": "pink",
    }
  }
>
  <ButtonCircle
    className="md-add-reaction-button-wrapper"
    size={20}
    style={
      {
        "color": "pink",
      }
    }
  >
    <ButtonSimple
      aria-disabled={false}
      className="md-button-circle-wrapper md-add-reaction-button-wrapper"
      data-color="primary"
      data-disabled={false}
      data-ghost={false}
      data-inverted={false}
      data-multiple-children={false}
      data-outline={false}
      data-shallow-disabled={false}
      data-size={20}
      style={
        {
          "color": "pink",
        }
      }
    >
      <FocusRing>
        <FocusRing
          focusClass="md-focus-ring-wrapper children"
          focusRingClass="md-focus-ring-wrapper children"
        >
          <button
            className="md-button-circle-wrapper md-add-reaction-button-wrapper md-button-simple-wrapper"
            data-color="primary"
            data-disabled={false}
            data-ghost={false}
            data-inverted={false}
            data-multiple-children={false}
            data-outline={false}
            data-shallow-disabled={false}
            data-size={20}
            onBlur={[Function]}
            onClick={[Function]}
            onDragStart={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchCancel={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            style={
              {
                "color": "pink",
              }
            }
            type="button"
          >
            <Icon
              name="reactions"
              scale={12}
            >
              <div
                aria-hidden="true"
                className="md-icon-wrapper md-icon-auto-scales md-icon-scales md-icon-no-shrink"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  className=""
                  data-autoscale={false}
                  data-scale={12}
                  data-test="reactions"
                  fill="currentColor"
                  height="100%"
                  style={{}}
                  viewBox="0, 0, 32, 32"
                  width="100%"
                />
              </div>
            </Icon>
          </button>
        </FocusRing>
      </FocusRing>
    </ButtonSimple>
  </ButtonCircle>
</AddReactionButton>
`;
